### PR TITLE
multi-interface network support

### PR DESCRIPTION
**Pull request by nickethier**

### CODE REVIEW

**Comment by shoenig** [comment]
This looks great, @nickethier !

Seems like we might want to come up with a test matrix for the upgrade path, that we can at least go through manually before GA. It's hard for me to think about exactly what's going to happen between the old and new structs

File: client/fingerprint/network_test.go [hunk 1]
```diff
@@ -197,6 +198,8 @@ func TestNetworkFingerprint_basic(t *testing.T) {
 		t.Fatalf("err: %v", err)
 	}
 
+	spew.Dump(response)
+	os.Exit(0)
```
Comment (schmichael) [inline client/fingerprint/network_test.go:202]: Left in some debugging code

File: client/taskenv/env.go [hunk 2]
```diff
@@ -70,15 +70,21 @@ const (
 	// The ip:port are always the host's.
 	AddrPrefix = "NOMAD_ADDR_"
 
+	HostAddrPrefix = "NOMAD_HOST_ADDR_"
```
Comment (schmichael) [inline client/taskenv/env.go:73]: All of these need comments.

File: go.mod [hunk 3]
```diff
@@ -36,6 +36,7 @@ require (
 	github.com/coreos/go-semver v0.3.0
 	github.com/coreos/pkg v0.0.0-20180928190104-399ea9e2e55f // indirect
 	github.com/cyphar/filepath-securejoin v0.2.3-0.20190205144030-7efe413b52e1 // indirect
+	github.com/davecgh/go-spew v1.1.1
```
Comment (schmichael) [inline go.mod:39]: Also due to debugging code above
```suggestion
```

File: nomad/structs/network.go [hunk 4]
```diff
@@ -226,6 +261,30 @@ func (idx *NetworkIndex) AddReservedPortRange(ports string) (collide bool) {
 	return
 }
 
+// AddReservedPortsForIP
```
Comment (schmichael) [inline nomad/structs/network.go:264]: Finish + comment other funcs

File: nomad/structs/network_test.go [hunk 5]
```diff
@@ -10,14 +10,15 @@ import (
 )
 
 func TestNetworkIndex_Overcommitted(t *testing.T) {
+	t.Skip()
```
Comment (schmichael) [inline nomad/structs/network_test.go:13]: Remove?

File: scheduler/generic_sched.go [hunk 6]
```diff
@@ -500,6 +500,7 @@ func (s *GenericScheduler) computePlacements(destructive, place []placementResul
 				}
 				if option.AllocResources != nil {
 					resources.Shared.Networks = option.AllocResources.Networks
+					resources.Shared.Ports = option.AllocResources.Ports
```
Comment (schmichael) [inline scheduler/generic_sched.go:503]: Need to make this change in the system scheduler too

**Pull request merged**